### ISSUE TITLE
password authentication

### ISSUE THREAD

**Issue by Alex9025** [Kitware/visualizer#56]
Hello all,

I tried to set up password authentication for Visualizer by Apache, with the configuration :

> <Directory "/var/web/visualizer/www"> 
>  AuthType Basic 
>  AuthName "Restricted Content" 
>  AuthUserFile /etc/apache2/.htpasswd 
> Require valid-user
> < /Directory>

Sometimes it works, but sometimes no authentication is required. Can someone advise me what should I do? 

Thanks!

Alex

**Comment by jourdain** [comment]
I believe that the part you need to protect are:
- /proxy
- /paraview

But the password will have to be requested before those endpoint are requested.

But I've never configured an apache to achieve such configuration, so I'm not sure I'll be of much help.

**Comment by Alex9025** [comment]
Thank you Sebastien. 

Can the option "--authKey", "${secret}"  restrict access ?

**Comment by jourdain** [comment]
Yes it will prevent any one to connect to a running session without that key.
That key can be either computed on the server side or the client can provide it.

But the one the client want to use needs to match the one used to start the pvpython process.

**Comment by Alex9025** [comment]
That's great! But I do not know how to do it, are there examples?

**Comment by jourdain** [comment]
I think you are getting confused on what it is meant for. And there are several option/path to support what you are trying to do, but usually that path require some custom deployment. With just apache+launcher and the default visualizer page, you will run into limitation. 

But here is an example below on how things could work.

The object that is passed as argument [here](https://github.com/Kitware/visualizer/blob/master/dist/index.html#L14) can be automatically extended with argument from url.

So for instance if you connect to visualizer with the following url args... `?secret=mySuperSecret&user=seb`

On the launcher side, if your command line looks like:

```
cmd: [
   'PASSWD', '=', 'getUserPassword(${user})', '&&', ..., '--authKey', '$PASSWD'
],
```

Then that user will only be able to connect to the server process only if the value PASSWD generated on the server machine match "mySuperSecret" string.

In similar way, you can fix the argument next to --authKey and only users that know the value will be able to connect if they provide it either from the URL or if the generated html page that start Visualizer have the correct value.

**Comment by Alex9025** [comment]
Thank you!
The second way with the key in the URL is simple and works very well. Is there a html page to let users enter the key?
The first way seems better, it allows to define a key by user, for this I have to reconfigure launcher, ... , it would be helpful to have more docs and examples on this topic.

**Comment by jourdain** [comment]
You need to provide another landing page and at submit you just use the visualizer html page as destination using the GET action so they will be put in the URL.

The tricky part for the documentation on that topic is the degree of freedom you have for such setup.
The launcher documentation do provide the basics of what I was explaining above.

**Comment by Alex9025** [comment]
Big thanks!
